bundles/binding/org.openhab.binding.powermax/src/main/java/org/openhab/binding/powermax/internal/connector/PowerMaxReaderThread.java

```diff
@@ -0,0 +1,182 @@
+/**
+ * Copyright (c) 2010-2016, openHAB.org and others.
+ *
+ * All rights reserved. This program and the accompanying materials
+ * are made available under the terms of the Eclipse Public License v1.0
+ * which accompanies this distribution, and is available at
+ * http://www.eclipse.org/legal/epl-v10.html
+ */
+package org.openhab.binding.powermax.internal.connector;
+
+import java.io.IOException;
+import java.io.InputStream;
+import java.io.InterruptedIOException;
+import java.util.Arrays;
+
+import javax.xml.bind.DatatypeConverter;
+
+import org.openhab.binding.powermax.internal.message.PowerMaxCommDriver;
+import org.openhab.binding.powermax.internal.message.PowerMaxReceiveType;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * A class that reads messages from the Visonic alarm panel in a dedicated thread
+ *
+ * @author lolodomo
+ * @since 1.9.0
+ */
+
+public class PowerMaxReaderThread extends Thread {
+
+    private static final Logger logger = LoggerFactory.getLogger(PowerMaxReaderThread.class);
+
+    private boolean interrupted = false;
+    private InputStream in;
+    private PowerMaxConnector connector;
+
+    /**
+     * Constructor
+     *
+     * @param in
+     *            the input stream
+     * @param connector
+     *            the object that should handle the received message
+     */
+    public PowerMaxReaderThread(InputStream in, PowerMaxConnector connector) {
+        this.in = in;
+        this.connector = connector;
+    }
+
+    @Override
+    public void interrupt() {
+        interrupted = true;
+        super.interrupt();
+        try {
+            in.close();
+        } catch (IOException e) {
+        } // quietly close
+    }
+
+    @Override
+    public void run() {
+        logger.debug("Data listener started");
+
+        final int dataBufferMaxLen = 0xC0;
+
+        byte[] dataBuffer = new byte[dataBufferMaxLen];
+        int index = 0;
+        int msgLen = 0;
+        boolean variableLen = false;
+        byte[] logData;
+
+        try {
+
+            byte[] tmpData = new byte[20];
+            int len = -1;
+
+            while (((len = in.read(tmpData)) > 0) && !interrupted) {
+
+                for (int i = 0; i < len; i++) {
+
+                    if (index >= dataBufferMaxLen) {
+                        // too many bytes received, try to find new start
+                        logData = Arrays.copyOf(dataBuffer, index);
+                        logger.debug("Truncating message {}", DatatypeConverter.printHexBinary(logData));
```

Comment:
Note the overhead of `DatatypeConverter.printHexBinary(logData)` even when not in DEBUG logging level.